stdlib/public/Concurrency/Actor.cpp

```diff
@@ -1386,15 +1434,31 @@ void DefaultActorImpl::enqueue(Job *job, JobPriority priority) {
     if (needsScheduling || needsStealer)
       taskExecutor = TaskExecutorRef::fromTaskExecutorPreference(job);
 
+    // In some cases (we aren't scheduling the actor and priorities don't
+    // match) then we need to access `this` after the enqueue. But the enqueue
+    // can cause the job to run and release `this`, so we need to retain `this`
+    // in those cases. The conditional here matches the conditions where we can
+    // get to the code below that uses `this`.
+    // If the priorities don't match then we need access to `this` after the
+    // compare_exchange. But enqueueing this job may cause `this` to be
+    // concurrently released. Defensively retain it.
+    if (oldState.isScheduled() || !newState.isScheduled())
```

Comment:
Do we mean to nest this?

Why are we checking oldState.isScheduled? Shouldn't this just be `if newState.isScheduled()`?

Reply:
```
if (newState.isScheduled() || (oldState.getMaxPriority() != newState.getMaxPriority())) {
    thisRetainHelper.defensiveRetain()
}
```

Reply:
There's this early return which avoids touching `this`:

```
      if (!oldState.isScheduled() && newState.isScheduled()) {
        // We took responsibility to schedule the actor for the first time. See
        // also ownership rule (1)
        return scheduleActorProcessJob(newState.getMaxPriority(), taskExecutor);
      }
```

And then after that, there's this conditional which goes into the stuff which uses `this`:

```
      if (oldState.getMaxPriority() != newState.getMaxPriority()) {
```

So we only need the retain if the first condition is not true and the second condition is true, which is (or should be) what I wrote here. I nested them just because it seemed easier to understand than a big compound conditional but I can change that if you find it less clear.

Reply:
Ah got it. My brain finds it easier to parse the snippet below cause the state change between old vs new, is clearer. 

```
if ((!oldState.isScheduled() && newState.isScheduled()) || // We took responsibility for scheduling, take +1
    (oldState.getMaxPriority() != newState.getMaxPriority())) { // We need to escalate, take +1
} 
```

Reply:
I'm going to break it into two separate `bool` variables and then check them, that way the components get names and the conditionals can exactly match what's done below. Will update that shortly.

Reply:
Updated, now it's:

```
    bool willSchedule = !oldState.isScheduled() && newState.isScheduled();
    bool priorityMismatch = oldState.getMaxPriority() != newState.getMaxPriority();
    if (!willSchedule && priorityMismatch)
        thisRetainHelper.defensiveRetain();
```